rust/core-lib/src/editor.rs

```diff
@@ -712,6 +717,62 @@ impl Editor {
         */
     }
 
+    fn indent(&mut self) {
+        self.modify_indent(IndentDirection::In);
+    }
+
+    fn outdent(&mut self) {
+        self.modify_indent(IndentDirection::Out);
+    }
+
+    /// Indents lines based on selection, while taking care to preserve cursor position as much as possible.
+    /// Outdent checks if each line in the selection is already fully tabbed (that is, larger than the config's tab size),
+    /// and deletes a tab character at start of line if it is. Otherwise, it checks for the first non-whitespace character and deletes 
+    /// up to that character instead.
+    fn modify_indent(&mut self, direction: IndentDirection) {
+        let mut builder = delta::Builder::new(self.text.len());
+        let tab_text = if self.config.items.translate_tabs_to_spaces {
+                    let tab_size = self.config.items.tab_size;
+                    n_spaces(tab_size)
+            } else {
+                "\t"
+            };
+        for region in self.view.sel_regions() {
+            let (first_line, _) = self.view.offset_to_line_col(&self.text, region.min());
+            let (last_line, last_col) =
+                self.view.offset_to_line_col(&self.text, region.max());
+            let last_line = if last_col == 0 && last_line > first_line {
+                last_line
+            } else {
+                last_line + 1
+            };    
+            for line in first_line..last_line {
+                let offset = self.view.offset_of_line(&self.text, line);
+                match direction {
+                    IndentDirection::In => {
+                        let interval = Interval::new_closed_open(offset, offset);  
+                        builder.replace(interval, Rope::from(tab_text));
+                        self.this_edit_type = EditType::InsertChars;
+                    },
+                    IndentDirection::Out => {
```

Comment:
this is also starting to drift pretty far to the right. This will be quite a better better if you end up just having a collection of lines and you iterate them, but I might also consider just adding `indent_line` and `outdent_line` fns that take a line number and contain the indent/outdent logic?